src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivitiesService.kt

```diff
@@ -74,9 +76,55 @@ class ActivitiesService(
       )
       log.debug("OffenderProgramProfile created with Nomis id = $offenderProgramReferenceId")
 
-      CreateOffenderProgramProfileResponse(offenderProgramReferenceId)
+      OffenderProgramProfileResponse(offenderProgramReferenceId)
     }
 
+  fun endOffenderProgramProfile(
+    courseActivityId: Long,
+    bookingId: Long,
+    dto: EndOffenderProgramProfileRequest
+  ): OffenderProgramProfileResponse =
+
+    offenderProgramProfileRepository.findByCourseActivityIdAndOffenderBookingIdAndAlloc(courseActivityId, bookingId)
+      ?.run {
+
+        dto.endReason?.also {
+          programServiceEndReasonRepository.findByIdOrNull(ProgramServiceEndReason.pk(it))?.also {
+            this.endReason = it
+          }
+            ?: throw BadDataException("End reason code=${dto.endReason} is invalid")
+        }
+        endDate = dto.endDate
+        endComment = dto.endComment
+
+        telemetryClient.trackEvent(
+          "offender-program-profile-ended",
+          mapOf(
+            "id" to offenderProgramReferenceId.toString(),
+            "courseActivityId" to courseActivity?.courseActivityId.toString(),
+            "bookingId" to offenderBooking.bookingId.toString(),
+          ),
+          null
+        )
+        log.debug("OffenderProgramProfile with Nomis id = $offenderProgramReferenceId set as ended")
```

Comment:
Are you planning to remove all these debug logs later? I don't think we need to double up on logs and telemetry.

Reply:
good point - removed